### PR TITLE
[ML] Integrating ML with the node shutdown API

### PR DESCRIPTION
**Pull request by droberts195**
When the node shutdown API is used to indicate that
an ML node is going to be shut down, ML will do the
following:

1. Isolate all datafeeds on the node and unassign
   their persistent tasks.
2. Tell all anomaly detection jobs on the node to go
   through the motions of closing, but not actually
   close and instead unassign their persistent tasks.
3. Report that the node is safe to shut down once all
   persistent tasks associated with anomaly detection
   jobs and model snapshot upgrades have either
   completed or been unassigned.

### CODE REVIEW

**Review by benwtrent — COMMENTED** [review]
Some of these interactions are fairly complex, but running through scenarios, there is only one that seems a tad strange to me. Stopping a vacating job.

Are we expecting that calling `_stop` against a job that is currently vacating will ALWAYS stop the job?

If so, it seems prudent to add a test that calls node shutdown, then `_stop` against a running job and then busy asserts to make sure the job gets stopped.

**Comment by droberts195** [comment]
> Are we expecting that calling `_stop` against a job that is currently vacating will ALWAYS stop the job?

`s/_stop/_close/` but yes, good catch.  At the moment if the close request gets redirected to the node that the job vacated, but slightly too late for the vacate to be turned back into a close then it gets lost.  The job will get reassigned to a different node instead of being closed.

In this situation I will make the close request that was redirected slightly too late delete the persistent task.  This will force-close the job on its new node.  This will be OK as it won't have been running long enough to have generated much in the way of new results or state.  Reverting to the snapshot that the vacate operation persisted will be fine in this case.

> If so, it seems prudent to add a test that calls node shutdown, then `_stop` against a running job and then busy asserts to make sure the job gets stopped.

Sure, I'll add a test when I make this scenario work.

**Comment by droberts195** [comment]
> In this situation I will make the close request that was redirected slightly too late delete the persistent task.

@benwtrent doing this opened up a nice can of worms: it turns out we _always_ had a problem if a job disappeared from the node we were waiting for it to close on while we were waiting.  It's just that the only way that could previously have happened is if somebody killed the node part way through the close operation and that's extremely difficult to test.

Anyway, the most recent commit seems to fix both that pre-existing loophole plus the case of closing a job while it's vacating the node.  On Monday I will run the two internal clusters tests added in this PR 100 times on my laptop to check they won't fail a high proportion of builds as soon as this PR is merged.

File: x-pack/plugin/ml/src/internalClusterTest/java/org/elasticsearch/xpack/ml/integration/MlNodeShutdownIT.java [hunk 1]
```diff
@@ -100,6 +101,97 @@ public void testJobsVacateShuttingDownNode() throws Exception {
...
+
+        client().execute(PutShutdownNodeAction.INSTANCE,
+            new PutShutdownNodeAction.Request(nodeIdToShutdown.get(), randomFrom(SingleNodeShutdownMetadata.Type.values()), "just testing"))
+            .actionGet();
+
+        if (randomBoolean()) {
+            // This isn't waiting for something to happen - just adding timing variation
+            // to increase the chance of subtle race conditions occurring.
+            Thread.sleep(randomIntBetween(1, 10));
+        }
```
Comment (benwtrent) [inline x-pack/plugin/ml/src/internalClusterTest/java/org/elasticsearch/xpack/ml/integration/MlNodeShutdownIT.java:164]: This may be good, but it seems to me that CI slowness/fastness might cause race conditions to bubble up themselves :)

File: x-pack/plugin/ml/src/internalClusterTest/java/org/elasticsearch/xpack/ml/integration/MlNodeShutdownIT.java [hunk 2]
```diff
@@ -100,6 +101,97 @@ public void testJobsVacateShuttingDownNode() throws Exception {
...
+            GetJobsStatsAction.Response statsResponse =
+                client().execute(GetJobsStatsAction.INSTANCE, new GetJobsStatsAction.Request(Metadata.ALL)).actionGet();
+            QueryPage<GetJobsStatsAction.Response.JobStats> jobStats = statsResponse.getResponse();
+            assertThat(jobStats, notNullValue());
+            long numJobsOnNodeToShutdown = jobStats.results().stream()
+                .filter(stats -> stats.getNode() != null && nodeNameToShutdown.equals(stats.getNode().getName())).count();
+            long numJobsOnOtherNodes = jobStats.results().stream()
+                .filter(stats -> stats.getNode() != null && nodeNameToShutdown.equals(stats.getNode().getName()) == false).count();
+            assertThat(numJobsOnNodeToShutdown, is(2L));
+            assertThat(numJobsOnOtherNodes, is(4L));
```
Comment (benwtrent) [inline x-pack/plugin/ml/src/internalClusterTest/java/org/elasticsearch/xpack/ml/integration/MlNodeShutdownIT.java:147]: Is it conceivable that things move SO fast that the jobs all get unassigned too quick so that this assertion always fails. I think assertBusy has an exponential backoff, so it seems to me that this could cause the test to be flaky.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportStopDatafeedAction.java [hunk 3]
```diff
@@ -363,16 +379,33 @@ private void sendResponseOrFailure(String datafeedId, ActionListener<StopDatafee
...
-    // so wait for that to happen here.
-    void waitForDatafeedStopped(List<String> datafeedPersistentTaskIds, StopDatafeedAction.Request request,
+    /**
+     * Wait for datafeed to be marked as stopped in cluster state, which means the datafeed persistent task has been removed.
+     * This api returns when task has been cancelled, but that doesn't mean the persistent task has been removed from cluster state,
+     * so wait for that to happen here.
+     *
+     * Since the stop datafeed action consists of a chain of async callbacks, it's possible that datafeeds have moved nodes since we
+     * decided what to do with them at the beginning of the chain.  We cannot simply wait for these, as the request to stop them will
+     * have been sent to the wrong node and ignored there, so we'll just spin until the timeout expires.
```
Comment (benwtrent) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportStopDatafeedAction.java:389]: Excellent catch here. Yeah, this has always been a weird edge case. Glad its found now :)

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportStopDatafeedAction.java [hunk 4]
```diff
@@ -194,21 +195,36 @@ private void normalStopDatafeed(Task task, StopDatafeedAction.Request request, A
...
+
         request.setNodes(executorNodes.toArray(new String[0]));
 
-        // wait for started and stopping datafeeds
-        // Map datafeedId -> datafeed task Id.
-        List<String> allDataFeedsToWaitFor = Stream.concat(
-                startedDatafeeds.stream().map(MlTasks::datafeedTaskId),
-                stoppingDatafeeds.stream().map(MlTasks::datafeedTaskId))
-                .collect(Collectors.toList());
+        final Set<String> movedDatafeeds = new HashSet<>();
```
Comment (benwtrent) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportStopDatafeedAction.java:210]: Does this need to be a `ConcurrentHashSet` or whatever because the collection is periodically updated/read via another thread?

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportCloseJobAction.java [hunk 5]
```diff
@@ -524,39 +526,75 @@ private void normalCloseJob(ClusterState currentState, Task task, CloseJobAction
             return;
         }
 
+        final Set<String> movedJobs = new HashSet<>();
```
Comment (benwtrent) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportCloseJobAction.java:529]: Should this set be `Concurrent` since its written to in one thread and read from others?

File: x-pack/plugin/ml/src/internalClusterTest/java/org/elasticsearch/xpack/ml/integration/MlNodeShutdownIT.java [hunk 6]
```diff
@@ -100,6 +101,97 @@ public void testJobsVacateShuttingDownNode() throws Exception {
...
+        });
+
+        client().execute(PutShutdownNodeAction.INSTANCE,
+            new PutShutdownNodeAction.Request(nodeIdToShutdown.get(), randomFrom(SingleNodeShutdownMetadata.Type.values()), "just testing"))
+            .actionGet();
+
+        if (randomBoolean()) {
+            // This isn't waiting for something to happen - just adding timing variation
+            // to increase the chance of subtle race conditions occurring.
+            Thread.sleep(randomIntBetween(1, 10));
```
Comment (benwtrent) [inline x-pack/plugin/ml/src/internalClusterTest/java/org/elasticsearch/xpack/ml/integration/MlNodeShutdownIT.java:163]: This may be good for local runs, but it seems to me that CI slowness/fastness might cause race conditions to bubble up themselves :)

File: x-pack/plugin/ml/src/internalClusterTest/java/org/elasticsearch/xpack/ml/integration/MlNodeShutdownIT.java (hunk 2 above)
Comment (benwtrent) [inline x-pack/plugin/ml/src/internalClusterTest/java/org/elasticsearch/xpack/ml/integration/MlNodeShutdownIT.java:147]: Is it conceivable that things move SO fast that the jobs all get unassigned too quick so that this assertion always fails. I think assertBusy has an exponential backoff, so it seems to me that this could cause the test to be flaky.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportStopDatafeedAction.java (hunk 4 above)
Comment (droberts195) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportStopDatafeedAction.java:210]: I _think_ it should be OK as-is, because all the insertions are done in the cluster state listener thread, and then that will also call the listener where the contents are used.  However, since it will be very rare that anything gets put in this map it doesn't really hurt to make it a concurrent set just in case the usage pattern changes in the future.

File: x-pack/plugin/ml/src/internalClusterTest/java/org/elasticsearch/xpack/ml/integration/MlNodeShutdownIT.java (hunk 2 above)
Comment (droberts195) [reply x-pack/plugin/ml/src/internalClusterTest/java/org/elasticsearch/xpack/ml/integration/MlNodeShutdownIT.java:135]: The shutdown API is called after this `assertBusy` block, not before.  I think my comments might have been confusing because I had a comment next to the declaration of the variables that remember which node is being shut down saying "Call the shutdown API" so I have made the comments more accurate about which bits do what.

There's a complication because the internal cluster test functionality works in terms of node names and the shutdown API works in terms of node IDs, so we have to find the node ID corresponding to a node name.  Since the ML stats API coincidentally returns that information for nodes where ML jobs are running I got it from the output of that call rather than making a separate call.

File: server/src/main/java/org/elasticsearch/persistent/PersistentTasksClusterService.java [hunk 7]
```diff
@@ -317,7 +317,8 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
         Randomness.shuffle(candidateNodes);
 
         final Assignment assignment = persistentTasksExecutor.getAssignment(taskParams, candidateNodes, currentState);
-        assert (assignment == null || isNodeShuttingDown(currentState, assignment.getExecutorNode()) == false) :
+        assert assignment != null : "getAssignment() should always return an Assignment object, containing a node or a reason why not";
+        assert (assignment.getExecutorNode() == null || isNodeShuttingDown(currentState, assignment.getExecutorNode()) == false) :
```
Comment (droberts195) [inline server/src/main/java/org/elasticsearch/persistent/PersistentTasksClusterService.java:321]: @dakrone and @henningandersen you may wish to review this change (just these 3 lines not the whole PR), as this change is in core code related to persistent tasks and the node shutdown API.

I am pretty sure that the idea was that `getAssignment` never returns `null`, but shout if you disagree.

I was getting NPEs from the original assertion when `assignment` was not `null` but `assignment.getExecutorNode()` was `null`.

**Comment by droberts195** [comment]
The CI failure in https://gradle-enterprise.elastic.co/s/v6gy7snyfysni indicates a genuine problem.  It's probably another thing that has existed for a long time but never had test coverage before.

It is possible for the post-test cleanup to fail to stop the datafeed that's just moved to a different node when the test ends - in this case it's `shutdown-job-2`:

```
[2021-07-19T19:04:31,348][INFO ][o.e.x.m.j.p.a.AutodetectProcessManager] [node_t0] Successfully set job state to [opened] for job [shutdown-job-2]
[2021-07-19T19:04:31,383][INFO ][o.e.x.m.i.MlNodeShutdownIT] [testJobsVacateShuttingDownNode] [MlNodeShutdownIT#testJobsVacateShuttingDownNode]: Cleaning up datafeeds and jobs after test
[2021-07-19T19:04:31,393][INFO ][o.e.x.m.i.MlNodeShutdownIT] [testJobsVacateShuttingDownNode] Closing all datafeeds (using _all)
[2021-07-19T19:04:31,541][INFO ][o.e.x.m.d.DatafeedJob    ] [node_t2] [shutdown-job-6] Lookback has finished
[2021-07-19T19:04:31,548][INFO ][o.e.x.m.d.DatafeedRunner ] [node_t0] [stop_datafeed (api)] attempt to stop datafeed [shutdown-job-1] [2]
[2021-07-19T19:04:31,552][INFO ][o.e.x.m.d.DatafeedRunner ] [node_t0] [stop_datafeed (api)] attempt to stop datafeed [shutdown-job-1] for job [shutdown-job-1]
[2021-07-19T19:04:31,552][INFO ][o.e.x.m.d.DatafeedRunner ] [node_t0] [stop_datafeed (api)] try lock [5m] to stop datafeed [shutdown-job-1] for job [shutdown-job-1]...
[2021-07-19T19:04:31,552][INFO ][o.e.x.m.d.DatafeedRunner ] [node_t0] [stop_datafeed (api)] stopping datafeed [shutdown-job-1] for job [shutdown-job-1], acquired [true]...
[2021-07-19T19:04:31,553][INFO ][o.e.x.m.d.DatafeedRunner ] [node_t0] [stop_datafeed (api)] datafeed [shutdown-job-1] for job [shutdown-job-1] has been stopped
[2021-07-19T19:04:31,604][INFO ][o.e.x.m.d.DatafeedRunner ] [node_t0] [stop_datafeed (api)] attempt to stop datafeed [shutdown-job-4] [8]
[2021-07-19T19:04:31,604][INFO ][o.e.x.m.d.DatafeedRunner ] [node_t0] [stop_datafeed (api)] attempt to stop datafeed [shutdown-job-4] for job [shutdown-job-4]
[2021-07-19T19:04:31,605][INFO ][o.e.x.m.d.DatafeedRunner ] [node_t0] [stop_datafeed (api)] try lock [5m] to stop datafeed [shutdown-job-4] for job [shutdown-job-4]...
[2021-07-19T19:04:31,605][INFO ][o.e.x.m.d.DatafeedRunner ] [node_t0] [stop_datafeed (api)] stopping datafeed [shutdown-job-4] for job [shutdown-job-4], acquired [true]...
[2021-07-19T19:04:31,606][INFO ][o.e.x.m.d.DatafeedRunner ] [node_t0] [stop_datafeed (api)] datafeed [shutdown-job-4] for job [shutdown-job-4] has been stopped
[2021-07-19T19:04:31,684][INFO ][o.e.x.m.d.DatafeedRunner ] [node_t2] [stop_datafeed (api)] attempt to stop datafeed [shutdown-job-5] [10]
[2021-07-19T19:04:31,684][INFO ][o.e.x.m.d.DatafeedRunner ] [node_t2] [stop_datafeed (api)] attempt to stop datafeed [shutdown-job-5] for job [shutdown-job-5]
[2021-07-19T19:04:31,685][INFO ][o.e.x.m.d.DatafeedRunner ] [node_t2] [stop_datafeed (api)] try lock [5m] to stop datafeed [shutdown-job-5] for job [shutdown-job-5]...
[2021-07-19T19:04:31,685][INFO ][o.e.x.m.d.DatafeedRunner ] [node_t2] [stop_datafeed (api)] stopping datafeed [shutdown-job-5] for job [shutdown-job-5], acquired [true]...
[2021-07-19T19:04:31,686][INFO ][o.e.x.m.d.DatafeedRunner ] [node_t2] [stop_datafeed (api)] datafeed [shutdown-job-5] for job [shutdown-job-5] has been stopped
[2021-07-19T19:04:31,736][INFO ][o.e.x.m.d.DatafeedRunner ] [node_t2] [stop_datafeed (api)] attempt to stop datafeed [shutdown-job-6] [23]
[2021-07-19T19:04:31,736][INFO ][o.e.x.m.d.DatafeedRunner ] [node_t2] [stop_datafeed (api)] attempt to stop datafeed [shutdown-job-6] for job [shutdown-job-6]
[2021-07-19T19:04:31,737][INFO ][o.e.x.m.d.DatafeedRunner ] [node_t2] [stop_datafeed (api)] try lock [5m] to stop datafeed [shutdown-job-6] for job [shutdown-job-6]...
[2021-07-19T19:04:31,737][INFO ][o.e.x.m.d.DatafeedRunner ] [node_t2] [stop_datafeed (api)] stopping datafeed [shutdown-job-6] for job [shutdown-job-6], acquired [true]...
[2021-07-19T19:04:31,738][INFO ][o.e.x.m.d.DatafeedRunner ] [node_t2] [stop_datafeed (api)] datafeed [shutdown-job-6] for job [shutdown-job-6] has been stopped
[2021-07-19T19:04:31,849][INFO ][o.e.x.m.d.DatafeedRunner ] [node_t2] [stop_datafeed (api)] attempt to stop datafeed [shutdown-job-3] [6]
[2021-07-19T19:04:31,849][INFO ][o.e.x.m.d.DatafeedRunner ] [node_t2] [stop_datafeed (api)] attempt to stop datafeed [shutdown-job-3] for job [shutdown-job-3]
[2021-07-19T19:04:31,849][INFO ][o.e.x.m.d.DatafeedRunner ] [node_t2] [stop_datafeed (api)] try lock [5m] to stop datafeed [shutdown-job-3] for job [shutdown-job-3]...
[2021-07-19T19:04:31,849][INFO ][o.e.x.m.d.DatafeedRunner ] [node_t2] [stop_datafeed (api)] stopping datafeed [shutdown-job-3] for job [shutdown-job-3], acquired [true]...
[2021-07-19T19:04:31,851][INFO ][o.e.x.m.d.DatafeedRunner ] [node_t2] [stop_datafeed (api)] datafeed [shutdown-job-3] for job [shutdown-job-3] has been stopped
[2021-07-19T19:04:31,905][INFO ][o.e.x.m.d.DatafeedJob    ] [node_t0] [shutdown-job-2] Datafeed started (from: 1970-01-01T00:00:00.000Z to: real-time) with frequency [1000ms]
[2021-07-19T19:04:32,082][INFO ][o.e.x.m.d.DatafeedJob    ] [node_t0] [shutdown-job-2] Lookback has finished
```

So it looks like we have a bug where a `_stop` of `_all` datafeeds can miss one that's recently switched nodes.

**Review by davidkyle — APPROVED** [review]
LGTM. That was a lot more complicated than I expected.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportStartDatafeedAction.java [hunk 8]
```diff
@@ -530,20 +555,48 @@ public boolean shouldCancelChildrenOnCancellation() {
...
         public void stop(String reason, TimeValue timeout) {
-            if (datafeedRunner != null) {
-                datafeedRunner.stopDatafeed(this, reason, timeout);
+            synchronized (this) {
+                if (datafeedRunner == null) {
+                    stoppedOrIsolatedBeforeRunning = StoppedOrIsolatedBeforeRunning.STOPPED;
+                    return;
+                }
             }
+            datafeedRunner.stopDatafeed(this, reason, timeout);
```
Comment (benwtrent) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportStartDatafeedAction.java:564]: There are still race conditions inside of `datafeedRunner`

Assume `datafeedRunner.run` has been called, but `runningDatafeedsOnThisNode.put(task.getAllocationId(), holder);` has not yet been set.

If so, during this window between `datafeedRunner.run` and the holder being set, calling `stop` does not prevent the `holder` from being created, added to the table of running datafeeds, and then the task state being set to opened.

Basically `finishHandler` is never called from what I can tell if this occurs. We should do something similar to what we do with the Job task by having a boolean set when `stop` is called that way the thing that creates/sets the holder can stop the datafeed.

File: x-pack/plugin/ml/src/internalClusterTest/java/org/elasticsearch/xpack/ml/integration/MlNodeShutdownIT.java [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+
+        // Open 6 jobs.  Since there are 3 nodes in the cluster we should get 2 jobs per node.
+        setupJobAndDatafeed("shutdown-close-job-1", ByteSizeValue.ofMb(2));
+        setupJobAndDatafeed("shutdown-close-job-2", ByteSizeValue.ofMb(2));
+        setupJobAndDatafeed("shutdown-close-job-3", ByteSizeValue.ofMb(2));
+        setupJobAndDatafeed("shutdown-close-job-4", ByteSizeValue.ofMb(2));
+        setupJobAndDatafeed("shutdown-close-job-5", ByteSizeValue.ofMb(2));
+        setupJobAndDatafeed("shutdown-close-job-6", ByteSizeValue.ofMb(2));
+
+        // Choose one non-master node to shut down, and one job on that node to close after the shutdown request has been sent.
```
[the remark this answers is not in the document]
Comment (droberts195) [reply x-pack/plugin/ml/src/internalClusterTest/java/org/elasticsearch/xpack/ml/integration/MlNodeShutdownIT.java:118]: The shutdown API is primarily for Cloud and in Cloud the ML nodes are never the master node.  So in production it's likely that ML jobs will only ever get vacated from non-master nodes.

(I also suspect vacating a node other than the current master is more likely to fail if there are race conditions as there's more lag between API calls when there are a mix of master node actions and non-master node actions.)

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/MlTasks.java [hunk 10]
```diff
@@ -250,6 +250,68 @@ public static DataFrameAnalyticsState getDataFrameAnalyticsState(@Nullable Persi
...
+        if (tasks == null) {
+            return Collections.emptyList();
+        }
+
+        return tasks.findTasks(JOB_SNAPSHOT_UPGRADE_TASK_NAME, task -> {
+            if (nodeId.equals(task.getExecutorNode()) == false) {
+                return false;
+            }
+            SnapshotUpgradeTaskState state = (SnapshotUpgradeTaskState) task.getState();
+            if (state == null) {
```
[the remark this answers is not in the document]
Comment (droberts195) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/MlTasks.java:309]: There are two objects involved: `task.getState()` and `state.getState()`.  The `state == null` check is against the persistent task state whereas the enum comparison on the last line is against the state field of the persistent task state.

I'll rename a variable to make it a bit clearer.

File: x-pack/plugin/ml/src/internalClusterTest/java/org/elasticsearch/xpack/ml/integration/MlNodeShutdownIT.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (davidkyle) [reply x-pack/plugin/ml/src/internalClusterTest/java/org/elasticsearch/xpack/ml/integration/MlNodeShutdownIT.java:118]: If the shutdown API can be called against a master node there is no harm adding coverage for that. Maybe use the `LuceneTestCase.rarely()` function to decide whether or not to shutdown the master

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportStartDatafeedAction.java (hunk 8 above)
Comment (droberts195) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportStartDatafeedAction.java:564]: I think https://github.com/elastic/elasticsearch/pull/75188/commits/3fca44a2003207e1f579ddb4d345ac13f9db0be0 should fix this.

Unlike the job there's no native process to kill so it's not quite a as complicated.  The datafeed doesn't start doing stuff until its persistent task successfully goes into the `started` state, and that shouldn't happen if the task has been removed, unassigned, or its allocation ID has changed.  If any of those things happens then the request to update the datafeed state to `started` will fail with a `ResourceNotFoundException`, and we can remove the entry from the `runningDatafeedsOnThisNode` map (which we should have been doing before in the code that reacts to `ResourceNotFoundException` - it was being leaked before).  Then everything related to the datafeed should be garbage-collected.

I have also added a couple of short-circuits to bail out of the creation process at an earlier point if we detect the datafeed has been stopped or isolated before it gets inserted into the map.

**Review by benwtrent — APPROVED** [review]

**Comment by droberts195** [comment]
😢 I was just going to click merge but a local run has revealed yet another problem.

I think it's when the datafeed is successfully started but cannot do anything because the job doesn't have an `autodetect` process yet.  So it's listening for cluster state updates to tell it when to actually start work.  Then it gets stopped, but before it can be fully stopped a cluster state update arrives that kicks it into action.  And then it's doing work in when everyone else thinks it's stopped.

I'll add a fix for this tomorrow.

**Comment by droberts195** [comment]
> I think it's when the datafeed is successfully started but cannot do anything because the job doesn't have an `autodetect` process yet.

That theory is wrong - that case is already covered.

The problem is that when we detect datafeeds that were moved when the stop request was sent to their node we just remove the persistent task to stop them on the new node.  However, this changes the semantics of a graceful stop - the stop request will no longer wait for the datafeed to finish its current cycle.  This causes a problem during test cleanup because we then move on to delete the source indices, which causes the datafeed to fail when its index disappears.

I think the solution in this case is to do what we were already doing in: https://github.com/elastic/elasticsearch/blob/108be9fe0eb7efa814d320aca4e8ce650080c210/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportStopDatafeedAction.java#L255-L260

**Comment by droberts195** [comment]
> I think the solution in this case is to do what we were already doing in

It turns out this approach doesn't work either.

After a datafeed moves nodes there is a time window when the master node sees that it's assigned to a new node but that new node has not yet had time to react to the cluster state update that assigned the datafeed, so doesn't have a `DatafeedTask` running on it.  If the request to stop arrives at the node when it's in this situation then the request is completely lost.  The node running the datafeed will soon notice the cluster state update and start it up.  And the master node will wait for ages for the stop request it sent to be actioned.

So it seems that the only approach that's robust so soon after a relocation is to remove the persistent task.  That then leaves the problem of how to ensure that the node where it might have been running has realised that it shouldn't be running any more.  I think the solution to that is to use the list tasks API.  So we need a way to know the ID of the local task that corresponds to a particular datafeed ID so that we can wait for it to stop.

**Comment by droberts195** [comment]
I have an idea why we don’t see this problem in most of our integration tests. In the external cluster tests we have a "wait for pending tasks" step during in the cleanup. This was supposed to just be waiting for async notifications to be indexed. But it will also be coincidentally waiting for orphaned local datafeed tasks to realise that their corresponding persistent tasks have disappeared. The internal cluster tests don’t currently have this "wait for pending tasks" in their cleanup before they delete indices. To get this PR moving I’ll just add it, because the underlying problem of orphaned datafeed local tasks continuing to run for a cycle after their persistent tasks disappear is no worse for our internal cluster tests than for all our external cluster tests that sneakily avoid the problem.

I will also open an issue to change the way we wait for jobs and datafeeds to be completed when we gracefully stop/close them. We should be waiting for both the persistent task and local task to complete. But since we have lived with the problem for years this change can wait for a followup.

**Review by benwtrent — APPROVED** [review]
I think this is a step in the right direction. We shall see if there are anymore gaps in our checks through CI. 

I also like the idea of using the task's API more often. I think both the `job/_close` and `datafeed/_stop` APIs could benefit.

**Comment by droberts195** [comment]
😭 it still doesn't work.  Our whole philosophy of asking jobs or datafeeds to stop gracefully by sending a message to the node we think they're running on is only robust if they are well-established on that node.  If they've just been assigned then the request can get lost and we wait until a timeout occurs.

The "I'm well established on this node" was supposed to be communicated by moving the job/datafeed state from opening/starting to opened/started.  But this doesn't work after a node reassignment, only when they are opened/started for the very first time.  We need to be making more use of `isStatusStale()` and doing things differently for tasks whose status is stale.

**Comment by droberts195** [comment]
> We need to be making more use of `isStatusStale()`

This just goes from bad to worse.  It turns out the datafeed state doesn't even have an `isStatusStale()` method like the job state (both for anomaly detection and data frame analytics) does.  And adding this functionality would mean changing the whole structure of the datafeed task state representation, so isn't viable now - it's just a single value rather than an object that we can add more information into.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportStopDatafeedAction.java [hunk 11]
```diff
@@ -393,13 +463,30 @@ void waitForDatafeedStopped(List<String> datafeedPersistentTaskIds, StopDatafeed
                 throw org.elasticsearch.ExceptionsHelper
                         .convertToElastic(failedNodeExceptions.get(0));
             } else {
-                // This can happen when the actual task in the node no longer exists,
-                // which means the datafeed(s) have already been stopped.
-                return new StopDatafeedAction.Response(true);
```
Comment (droberts195) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportStopDatafeedAction.java:398]: We were discarding significant information here.  This is where we could have detected the case of "persistent task implies local task is running on node X, no local task currently running on node X".

When we get here we don't know if the reason is that the local task hasn't yet started or that the local task has finished but the persistent task hasn't yet been removed.  In the first case we have to do something.  In the second we don't, but it doesn't hurt to retry.

I didn't add similar logic to the close job code.  For some reason we never seem to hit the problematic case with jobs.  This may be because job state has the concept of being stale, so we can detect recently relocated jobs by their staleness.  I don't want to add more complex code unnecessarily, but if we ever see the case of jobs closures timing out then something like what's now below may be the solution.

**Comment by droberts195** [comment]
@elasticmachine update branch

**Comment by dakrone** [comment]
@droberts195 have you seen this error? I'm seeing it on master quite a bit:

```
[2021-07-28T20:54:28,213][WARN ][o.e.s.PluginShutdownService] [javaRestTest-3] uncaught exception when notifying plugins of nodes [] shutdown
java.lang.NullPointerException: null
	at org.elasticsearch.xpack.ml.MachineLearning.signalShutdown(MachineLearning.java:1537) ~[?:?]
	at org.elasticsearch.shutdown.PluginShutdownService.signalShutdown(PluginShutdownService.java:97) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.shutdown.PluginShutdownService.clusterChanged(PluginShutdownService.java:106) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.cluster.service.ClusterApplierService.callClusterStateListener(ClusterApplierService.java:510) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.cluster.service.ClusterApplierService.callClusterStateListeners(ClusterApplierService.java:500) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.cluster.service.ClusterApplierService.applyChanges(ClusterApplierService.java:468) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.cluster.service.ClusterApplierService.runTask(ClusterApplierService.java:408) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.cluster.service.ClusterApplierService$UpdateTask.run(ClusterApplierService.java:152) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:678) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.runAndClean(PrioritizedEsThreadPoolExecutor.java:259) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.run(PrioritizedEsThreadPoolExecutor.java:222) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
	at java.lang.Thread.run(Thread.java:834) [?:?]
```

It's not causing any failures, but it does spam the logs a bit. Looks like `mlLifeCycleService.get()` is returning null sometimes?

**Comment by droberts195** [comment]
> Looks like `mlLifeCycleService.get()` is returning null sometimes?

Ah yes, that will be when the plugin is disabled.  I'll fix that now.

File: server/src/main/java/org/elasticsearch/persistent/PersistentTasksClusterService.java (hunk 7 above)
Comment (henningandersen) [reply server/src/main/java/org/elasticsearch/persistent/PersistentTasksClusterService.java:321]: That change LGTM (sorry for the delayed response).
